File: java/src/main/java/decorator/decorator/WelcomeDiscount.java
package decorator.decorator;

import decorator.Discount;

public class WelcomeDiscount implements Discount {
    public static final double RATE = 0.01;
    public static final String HISTORY = "웰컴 할인";
    private final Discount discount;

    public WelcomeDiscount(Discount discount) {
        this.discount = discount;
    }

    @Override
    public double amountAfterDiscount(double amount) {
        return discount.amountAfterDiscount(amount) * RATE;
    }

    @Override
    public String history() {
        return discount.history() + " " + HISTORY;
    }
}

File: kotlin/src/main/kotlin/decorator/decorator/WelcomeDiscount.kt
package decorator.decorator

import decorator.Discount

class WelcomeDiscount(private val discount: Discount): Discount {
    companion object {
        const val RATE = 0.01
        const val HISTORY = "웰컴 할이"
    }

    override fun amountAfterDiscount(amount: Double): Double {
        return discount.amountAfterDiscount(amount) * RATE
    }

    override fun history(): String {
        return "${discount.history()} $HISTORY"
    }
}